tensorflow_datasets/core/dataset_builder.py

```diff
@@ -298,6 +298,27 @@ def download_and_prepare(self, download_dir=None, download_config=None):
               self.name, self.version, self.version.tfds_version_to_prepare,
               available_to_prepare))
 
+    # Check if requested version is canonical version or max(self.versions)
+    # if not raises error as tfds not supports older versions of datasets.
+    new_versions = [str(v) for v in [self.canonical_version] + [max(self.versions)]]
+    if str(self._version) not in new_versions:
+      # if canonical and maximum versions are same then suggest only one version
+      if max(self.versions) != self.canonical_version:
+        raise AssertionError(
```

Comment:
Rather than duplicating three times the error message, can't you refactor to only raise the exception once ?

Reply:
Yes its good to have only once, I will make changes

Reply:
@Conchylicultor  i am not understanding what you are asking when you look on changes there is a check for `if self.version.tfds_version_to_prepare:` which raises exception and then there is a check which check if version is canonical or maximum and it raises error and there is two assertion errors  because when canonical and max version is same it should recommend only one of them version .

Reply:
* To merge the two raise here, you should use `set()` instead of `list` so that duplicate versions are merged.
* To merge this one with the `tfds_version_to_prepare `, the `tfds_version_to_prepare` check should be added to this error message if present, as `tfds_version_to_prepare ` is only present for older versions.

Reply:
@Conchylicultor  thanks for suggestions it is good to use set as they take only unique elements in it just forgot to use them here, I will make changes soon

Reply:
@Conchylicultor  if we merge `tfds_version_to_prepare` check with new version checks so they both show same error message , it is good to include unittest for  `tfds_version_to_prepare` as both of them shows same errors.

Reply:
Yes, we should have unitests for the different way to raise error.

Reply:
Thanks for reply I will make changes soon.

Reply:
@Conchylicultor  If we are including both unittests so I think I have to change in `test_utils.py file` in `utils.Version("0.0.7", tfds_version_to_prepare="v1.0.0"),]`, 0.0.7 to 2.0.0 for test passes. Also for test purpose you suggest to create new specific dataset in `test_with_old_version` so should I include mnist or re create new one in `test_util`